src/composer.ts

```diff
@@ -12,6 +13,34 @@ type Triggers<C> = MaybeArray<
 type Predicate<T> = (t: T) => boolean
 type AsyncPredicate<T> = (t: T) => Promise<boolean>
 
+type UnionToIntersection<U> = (
+  U extends unknown ? (k: U) => void : never
+) extends (k: infer I) => void
+  ? I
+  : never
+type PropOr<T, P extends string, D> = T extends Record<P, infer V> ? V : D
+type GuardedContext<
+  C extends Context,
+  U extends Omit<tt.Update, 'update_id'>
+> = C &
+  {
+    [P in tt.UpdateType as SnakeToCamelCase<P>]: PropOr<U, P, undefined>
+  } & {
+    update: U
+    updateType: keyof UnionToIntersection<U>
+  }
+type MountMap = {
+  [T in tt.UpdateType]: Record<T, object>
+} &
+  {
+    [T in tt.MessageSubType]: {
+      message: Record<
+        PropOr<typeof MessageSubTypesMapping, T, T>,
+        UnionToIntersection<tt.Message>[T]
+      >
+    }
+  }
+
```

Comment:
Should this be grouped together with `MatchedContext`, rather than being separated from it by the helper functions?

Reply:
If anything, `MatchedContext` and `MatchedMiddleware` are the only type aliases that are not grouped with the rest, and I don't wish to have `git blame` on those lines, so I'm leaving as-is :)

Reply:
I can move them for you if that's the issue. After all, they are heavily dependent on each other and belong semantically together, too.